File: src/main/java/jugglinglab/view/PatternView.java
//
// PatternView.java
//
// This view provides the ability to edit the text representation of a pattern.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.view;

import java.awt.*;
import java.io.File;
import java.io.StringReader;
import java.text.MessageFormat;
import javax.swing.*;
import javax.swing.event.*;
import jugglinglab.core.AnimationPanel;
import jugglinglab.core.AnimationPrefs;
import jugglinglab.jml.JMLPattern;
import jugglinglab.util.*;

public class PatternView extends View implements DocumentListener {
  protected AnimationPanel ja;
  protected JSplitPane jsp;
  protected JRadioButton rb_bp;
  protected JLabel bp_edited_icon;
  protected JRadioButton rb_jml;
  protected JTextArea ta;
  protected JButton compile;
  protected JButton revert;
  protected JLabel lab;
  protected boolean text_edited = false;

  public PatternView(Dimension dim) {
    makePanel(dim);
    updateButtons();
  }

  protected void makePanel(Dimension dim) {
    setLayout(new BorderLayout());

    // animator on the left
    ja = new AnimationPanel();
    ja.setPreferredSize(dim);
    ja.setMinimumSize(new Dimension(10, 10));

    // controls panel on the right
    JPanel controls = new JPanel();
    GridBagLayout gb = new GridBagLayout();
    controls.setLayout(gb);

    JLabel lab_view = new JLabel(guistrings.getString("PatternView_heading"));
    gb.setConstraints(
        lab_view,
        JLFunc.constraints(GridBagConstraints.LINE_START, 0, 0, new Insets(15, 4, 10, 0)));
    controls.add(lab_view);

    ButtonGroup bg = new ButtonGroup();
    JPanel bppanel = new JPanel();
    bppanel.setLayout(new FlowLayout(FlowLayout.LEFT, 0, 0));
    rb_bp = new JRadioButton(guistrings.getString("PatternView_rb1_default"));
    bg.add(rb_bp);
    bppanel.add(rb_bp);
    java.net.URL url = PatternView.class.getResource("/alert.png");
    if (url != null) {
      ImageIcon edited_icon = new ImageIcon(url);
      ImageIcon edited_icon_scaled =
          new ImageIcon(
              edited_icon.getImage().getScaledInstance(22, 22, Image.SCALE_SMOOTH));
      bp_edited_icon = new JLabel(edited_icon_scaled);
      bp_edited_icon.setToolTipText(guistrings.getString("PatternView_alert"));
      bppanel.add(Box.createHorizontalStrut(10));
      bppanel.add(bp_edited_icon);
    }
    controls.add(bppanel);
    gb.setConstraints(
        bppanel, JLFunc.constraints(GridBagConstraints.LINE_START, 0, 1, new Insets(0, 4, 0, 0)));

    rb_jml = new JRadioButton(guistrings.getString("PatternView_rb2"));
    bg.add(rb_jml);
    controls.add(rb_jml);
    gb.setConstraints(
        rb_jml, JLFunc.constraints(GridBagConstraints.LINE_START, 0, 2, new Insets(0, 4, 0, 0)));

    ta = new JTextArea();
    JScrollPane jscroll = new JScrollPane(ta);
    jscroll.setPreferredSize(new Dimension(400, 1));
    jscroll.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS);
    jscroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
    controls.add(jscroll);
    GridBagConstraints gbc =
        JLFunc.constraints(GridBagConstraints.LINE_START, 0, 3, new Insets(15, 0, 0, 0));
    gbc.fill = GridBagConstraints.BOTH;
    gbc.weightx = gbc.weighty = 1.0;
    gb.setConstraints(jscroll, gbc);

    // split pane dividing the two
    jsp = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, ja, controls);
    jsp.setResizeWeight(0.75); // % extra space allocated to left (animation) side

    add(jsp, BorderLayout.CENTER);

    // button + error message label across the bottom
    JPanel lower = new JPanel();
    GridBagLayout gb2 = new GridBagLayout();
    lower.setLayout(gb2);
    compile = new JButton(guistrings.getString("PatternView_compile_button"));
    gb2.setConstraints(
        compile, JLFunc.constraints(GridBagConstraints.LINE_START, 0, 0, new Insets(8, 8, 8, 0)));
    lower.add(compile);
    revert = new JButton(guistrings.getString("PatternView_revert_button"));
    gb2.setConstraints(
        revert, JLFunc.constraints(GridBagConstraints.LINE_START, 1, 0, new Insets(8, 5, 8, 12)));
    lower.add(revert);
    lab = new JLabel(" ");
    GridBagConstraints gbc2 = JLFunc.constraints(GridBagConstraints.LINE_START, 2, 0);
    gbc2.fill = GridBagConstraints.HORIZONTAL;
    gbc2.weightx = 1.0;
    gb2.setConstraints(lab, gbc2);
    lower.add(lab);

    add(lower, BorderLayout.PAGE_END);

    // add actions to the various items
    ta.getDocument().addDocumentListener(this);
    rb_bp.addActionListener(ae -> reloadTextArea());
    rb_jml.addActionListener(ae -> reloadTextArea());
    compile.addActionListener(ae -> compilePattern());
    revert.addActionListener(ae -> revertPattern());
  }

  // Update the button configs when a radio button is pressed, the base
  // pattern or JML pattern changes, or we start/stop writing an animated GIF.

  protected void updateButtons() {
    if (ja != null && ja.writingGIF) {
      // writing a GIF
      rb_bp.setEnabled(false);
      rb_jml.setEnabled(false);
      compile.setEnabled(false);
      revert.setEnabled(false);
      return;
    }

    JMLPattern pat = getPattern();

    if (pat == null) {
      rb_bp.setEnabled(false);
      if (bp_edited_icon != null) {
        bp_edited_icon.setVisible(false);
      }
      rb_jml.setEnabled(false);
    } else if (pat.getBasePatternNotation() == null || pat.getBasePatternConfig() == null) {
      // no base pattern set
      rb_bp.setEnabled(false);
      if (bp_edited_icon != null) {
        bp_edited_icon.setVisible(false);
      }
      rb_jml.setEnabled(true);
      rb_jml.setSelected(true);
    } else {
      rb_bp.setEnabled(true);
      if (bp_edited_icon != null) {
        bp_edited_icon.setVisible(pat.isBasePatternEdited());
      }
      rb_jml.setEnabled(true);
    }

    if (rb_bp.isSelected()) {
      compile.setEnabled(pat != null && (pat.isBasePatternEdited() || text_edited));
      revert.setEnabled(text_edited);
    } else if (rb_jml.isSelected()) {
      compile.setEnabled(text_edited);
      revert.setEnabled(text_edited);
    }
  }

  // (Re)load the text in the JTextArea from the pattern, overwriting anything
  // that was there.

  protected void reloadTextArea() {
    if (rb_bp.isSelected()) {
      ta.setText(getPattern().getBasePatternConfig().replace(";", ";\n"));
    } else if (rb_jml.isSelected()) {
      ta.setText(getPattern().toString());
    }

    ta.setCaretPosition(0);
    lab.setText(" ");
    setTextEdited(false);

    // Note the above always triggers an updateButtons() call, since
    // text_edited is cycled from true (from the setText() calls) to false.
  }

  protected void setTextEdited(boolean edited) {
    if (text_edited != edited) {
      text_edited = edited;
      updateButtons();
    }
  }

  protected void compilePattern() {
    try {
      if (rb_bp.isSelected()) {
        String notation = getPattern().getBasePatternNotation();
        String config = ta.getText().replace("\n", "").trim();
        JMLPattern newpat = JMLPattern.fromBasePattern(notation, config);
        restartView(newpat, null);
        addToUndoList(newpat);
      } else if (rb_jml.isSelected()) {
        JMLPattern newpat = new JMLPattern(new StringReader(ta.getText()));
        restartView(newpat, null);
        addToUndoList(newpat);
      }
    } catch (JuggleExceptionUser jeu) {
      lab.setText(jeu.getMessage());
      setTextEdited(true);
    } catch (JuggleExceptionInternal jei) {
      ErrorDialog.handleFatalException(jei);
      setTextEdited(true);
    }
  }

  protected void revertPattern() {
    reloadTextArea();
  }

  //----------------------------------------------------------------------------
  // View methods
  //----------------------------------------------------------------------------

  @Override
  public void restartView(JMLPattern p, AnimationPrefs c)
      throws JuggleExceptionUser, JuggleExceptionInternal {
    ja.restartJuggle(p, c);
    setAnimationPanelPreferredSize(getAnimationPrefs().getSize());

    if (p != null) {
      String notation = p.getBasePatternNotation();
      String template = guistrings.getString("PatternView_rb1");
      Object[] arg = {notation == null ? "none set" : notation};
      rb_bp.setText(MessageFormat.format(template, arg));

      if (!(rb_bp.isSelected() || rb_jml.isSelected())) {
        if (notation == null) {
          rb_jml.setSelected(true);
        } else {
          rb_bp.setSelected(true);
        }
      }

      updateButtons();
      reloadTextArea();
      parent.setTitle(p.getTitle());
    }
  }

  @Override
  public void restartView() throws JuggleExceptionUser, JuggleExceptionInternal {
    ja.restartJuggle();
  }

  @Override
  public Dimension getAnimationPanelSize() {
    return ja.getSize(new Dimension());
  }

  @Override
  public void setAnimationPanelPreferredSize(Dimension d) {
    ja.setPreferredSize(d);
    jsp.resetToPreferredSizes();
  }

  @Override
  public JMLPattern getPattern() {
    return ja.getPattern();
  }

  @Override
  public AnimationPrefs getAnimationPrefs() {
    return ja.getAnimationPrefs();
  }

  @Override
  public double getZoomLevel() {
    return ja.getZoomLevel();
  }

  @Override
  public void setZoomLevel(double z) {
    ja.setZoomLevel(z);
  }

  @Override
  public boolean isPaused() {
    return ja.isPaused();
  }

  @Override
  public void setPaused(boolean pause) {
    if (ja.message == null) {
      ja.setPaused(pause);
    }
  }

  @Override
  public void disposeView() {
    ja.disposeAnimation();
  }

  @Override
  public void writeGIF(File f) {
    ja.writingGIF = true;
    updateButtons();
    boolean origpause = isPaused();
    setPaused(true);
    jsp.setEnabled(false);
    if (parent != null) {
      parent.setResizable(false);
    }

    Runnable cleanup =
        () -> {
          ja.writingGIF = false;
          setPaused(origpause);
          updateButtons();
          jsp.setEnabled(true);
          if (parent != null) {
            parent.setResizable(true);
          }
        };

    new View.GIFWriter(ja, f, cleanup);
  }

  //----------------------------------------------------------------------------
  // javax.swing.event.DocumentListener methods
  //----------------------------------------------------------------------------

  @Override
  public void insertUpdate(DocumentEvent e) {
    setTextEdited(true);
  }

  @Override
  public void removeUpdate(DocumentEvent e) {
    setTextEdited(true);
  }

  @Override
  public void changedUpdate(DocumentEvent e) {
    setTextEdited(true);
  }
}

File: src/main/kotlin/jugglinglab/view/PatternView.kt
//
// PatternView.kt
//
// This view provides the ability to edit the text representation of a pattern.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.view

import jugglinglab.core.AnimationPanel
import jugglinglab.core.AnimationPrefs
import jugglinglab.jml.JMLPattern
import jugglinglab.JugglingLab.guistrings
import jugglinglab.util.ErrorDialog.handleFatalException
import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.constraints
import java.awt.*
import java.awt.event.ActionEvent
import java.io.File
import java.io.StringReader
import java.text.MessageFormat
import javax.swing.*
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener

class PatternView(dim: Dimension?) : View(), DocumentListener {
    private lateinit var ja: AnimationPanel
    private lateinit var jsp: JSplitPane
    private lateinit var rbBp: JRadioButton
    private var bpEditedIcon: JLabel? = null
    private lateinit var rbJml: JRadioButton
    private lateinit var ta: JTextArea
    private lateinit var compile: JButton
    private lateinit var revert: JButton
    private lateinit var lab: JLabel
    private var textEdited: Boolean = false

    init {
        makePanel(dim)
        updateButtons()
    }

    private fun makePanel(dim: Dimension?) {
        setLayout(BorderLayout())

        // animator on the left
        ja = AnimationPanel()
        ja.preferredSize = dim
        ja.minimumSize = Dimension(10, 10)

        // controls panel on the right
        val controls = JPanel()
        val gb = GridBagLayout()
        controls.setLayout(gb)

        val labView = JLabel(guistrings.getString("PatternView_heading"))
        gb.setConstraints(
            labView,
            constraints(GridBagConstraints.LINE_START, 0, 0, Insets(15, 4, 10, 0))
        )
        controls.add(labView)

        val bg = ButtonGroup()
        val bppanel = JPanel()
        bppanel.setLayout(FlowLayout(FlowLayout.LEFT, 0, 0))
        rbBp = JRadioButton(guistrings.getString("PatternView_rb1_default"))
        bg.add(rbBp)
        bppanel.add(rbBp)
        val url = PatternView::class.java.getResource("/alert.png")
        if (url != null) {
            val editedIcon = ImageIcon(url)
            val editedIconScaled =
                ImageIcon(
                    editedIcon.getImage().getScaledInstance(22, 22, Image.SCALE_SMOOTH)
                )
            bpEditedIcon = JLabel(editedIconScaled)
            bpEditedIcon?.setToolTipText(guistrings.getString("PatternView_alert"))
            bppanel.add(Box.createHorizontalStrut(10))
            bppanel.add(bpEditedIcon)
        }
        controls.add(bppanel)
        gb.setConstraints(
            bppanel, constraints(GridBagConstraints.LINE_START, 0, 1, Insets(0, 4, 0, 0))
        )

        rbJml = JRadioButton(guistrings.getString("PatternView_rb2"))
        bg.add(rbJml)
        controls.add(rbJml)
        gb.setConstraints(
            rbJml, constraints(GridBagConstraints.LINE_START, 0, 2, Insets(0, 4, 0, 0))
        )

        ta = JTextArea()
        val jscroll = JScrollPane(ta)
        jscroll.preferredSize = Dimension(400, 1)
        jscroll.setVerticalScrollBarPolicy(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS)
        jscroll.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS)
        controls.add(jscroll)
        val gbc =
            constraints(GridBagConstraints.LINE_START, 0, 3, Insets(15, 0, 0, 0))
        gbc.fill = GridBagConstraints.BOTH
        gbc.weighty = 1.0
        gbc.weightx = gbc.weighty
        gb.setConstraints(jscroll, gbc)

        // split pane dividing the two
        jsp = JSplitPane(JSplitPane.HORIZONTAL_SPLIT, true, ja, controls)
        jsp.setResizeWeight(0.75) // % extra space allocated to left (animation) side

        add(jsp, BorderLayout.CENTER)

        // button + error message label across the bottom
        val lower = JPanel()
        val gb2 = GridBagLayout()
        lower.setLayout(gb2)
        compile = JButton(guistrings.getString("PatternView_compile_button"))
        gb2.setConstraints(
            compile, constraints(GridBagConstraints.LINE_START, 0, 0, Insets(8, 8, 8, 0))
        )
        lower.add(compile)
        revert = JButton(guistrings.getString("PatternView_revert_button"))
        gb2.setConstraints(
            revert, constraints(GridBagConstraints.LINE_START, 1, 0, Insets(8, 5, 8, 12))
        )
        lower.add(revert)
        lab = JLabel(" ")
        val gbc2 = constraints(GridBagConstraints.LINE_START, 2, 0)
        gbc2.fill = GridBagConstraints.HORIZONTAL
        gbc2.weightx = 1.0
        gb2.setConstraints(lab, gbc2)
        lower.add(lab)

        add(lower, BorderLayout.PAGE_END)

        // add actions to the various items
        ta.document.addDocumentListener(this)
        rbBp.addActionListener { _: ActionEvent? -> reloadTextArea() }
        rbJml.addActionListener { _: ActionEvent? -> reloadTextArea() }
        compile.addActionListener { _: ActionEvent? -> compilePattern() }
        revert.addActionListener { _: ActionEvent? -> revertPattern() }
    }

    // Update the button configs when a radio button is pressed, the base
    // pattern or JML pattern changes, or we start/stop writing an animated GIF.
    private fun updateButtons() {
        if (ja.writingGIF) {
            // writing a GIF
            rbBp.setEnabled(false)
            rbJml.setEnabled(false)
            compile.setEnabled(false)
            revert.setEnabled(false)
            return
        }

        val pat = pattern

        if (pat == null) {
            rbBp.setEnabled(false)
            bpEditedIcon?.isVisible = false
            rbJml.setEnabled(false)
        } else if (pat.basePatternNotation == null || pat.basePatternConfig == null) {
            // no base pattern set
            rbBp.setEnabled(false)
            bpEditedIcon?.isVisible = false
            rbJml.setEnabled(true)
            rbJml.setSelected(true)
        } else {
            rbBp.setEnabled(true)
            bpEditedIcon?.isVisible = pat.isBasePatternEdited
            rbJml.setEnabled(true)
        }

        if (rbBp.isSelected) {
            compile.setEnabled(pat != null && (pat.isBasePatternEdited || textEdited))
            revert.setEnabled(textEdited)
        } else if (rbJml.isSelected) {
            compile.setEnabled(textEdited)
            revert.setEnabled(textEdited)
        }
    }

    // (Re)load the text in the JTextArea from the pattern, overwriting anything
    // that was there.
    private fun reloadTextArea() {
        if (rbBp.isSelected) {
            ta.text = pattern!!.basePatternConfig!!.replace(";", ";\n")
        } else if (rbJml.isSelected) {
            ta.text = pattern.toString()
        }

        ta.setCaretPosition(0)
        lab.setText(" ")
        setTextEdited(false)

        // Note the above always triggers an updateButtons() call, since
        // text_edited is cycled from true (from the setText() calls) to false.
    }

    private fun setTextEdited(edited: Boolean) {
        if (textEdited != edited) {
            textEdited = edited
            updateButtons()
        }
    }

    private fun compilePattern() {
        try {
            if (rbBp.isSelected) {
                val notation = pattern!!.basePatternNotation
                val config = ta.getText().replace("\n", "").trim { it <= ' ' }
                val newpat = JMLPattern.fromBasePattern(notation, config)
                restartView(newpat, null)
                addToUndoList(newpat)
            } else if (rbJml.isSelected) {
                val newpat = JMLPattern(StringReader(ta.getText()))
                restartView(newpat, null)
                addToUndoList(newpat)
            }
        } catch (jeu: JuggleExceptionUser) {
            lab.setText(jeu.message)
            setTextEdited(true)
        } catch (jei: JuggleExceptionInternal) {
            handleFatalException(jei)
            setTextEdited(true)
        }
    }

    private fun revertPattern() {
        reloadTextArea()
    }

    //--------------------------------------------------------------------------
    // View methods
    //--------------------------------------------------------------------------

    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    override fun restartView(p: JMLPattern?, c: AnimationPrefs?) {
        ja.restartJuggle(p, c)
        setAnimationPanelPreferredSize(animationPrefs.size)

        if (p != null) {
            val notation = p.basePatternNotation
            val template = guistrings.getString("PatternView_rb1")
            val arg = arrayOf<Any?>(notation ?: "none set")
            rbBp.setText(MessageFormat.format(template, *arg))

            if (!(rbBp.isSelected || rbJml.isSelected)) {
                if (notation == null) {
                    rbJml.setSelected(true)
                } else {
                    rbBp.setSelected(true)
                }
            }

            updateButtons()
            reloadTextArea()
            patternWindow?.setTitle(p.title)
        }
    }

    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    override fun restartView() {
        ja.restartJuggle()
    }

    override val animationPanelSize: Dimension?
        get() = ja.getSize(Dimension())

    override fun setAnimationPanelPreferredSize(d: Dimension) {
        ja.preferredSize = d
        jsp.resetToPreferredSizes()
    }

    override val pattern: JMLPattern?
        get() = ja.pattern

    override val animationPrefs: AnimationPrefs
        get() = ja.animationPrefs

    override var zoomLevel: Double
        get() = ja.zoomLevel
        set(z) {
            ja.zoomLevel = z
        }

    override var isPaused: Boolean
        get() = ja.isPaused
        set(pause) {
            if (ja.message == null) {
                ja.isPaused = pause
            }
        }

    override fun disposeView() {
        ja.disposeAnimation()
    }

    override fun writeGIF(f: File) {
        ja.writingGIF = true
        updateButtons()
        val origpause = isPaused
        isPaused = true
        jsp.isEnabled = false
        patternWindow?.isResizable = false

        val cleanup =
            Runnable {
                ja.writingGIF = false
                isPaused = origpause
                updateButtons()
                jsp.isEnabled = true
                patternWindow?.isResizable = true
            }

        GIFWriter(ja, f, cleanup)
    }

    //--------------------------------------------------------------------------
    // javax.swing.event.DocumentListener methods
    //--------------------------------------------------------------------------

    override fun insertUpdate(e: DocumentEvent?) {
        setTextEdited(true)
    }

    override fun removeUpdate(e: DocumentEvent?) {
        setTextEdited(true)
    }

    override fun changedUpdate(e: DocumentEvent?) {
        setTextEdited(true)
    }
}
